<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-McElroy2013">
<p>McElroy, Kerensa, Osvaldo Zagordi, Rowena Bull, Fabio Luciani, and Niko Beerenwinkel. 2013. “Accurate Single Nucleotide Variant Detection in Viral Populations by Combining Probabilistic Clustering with a Statistical Test of Strand Bias.” <em>BMC Genomics</em> 14 (1): 501. <a href="https://doi.org/10.1186/1471-2164-14-501">https://doi.org/10.1186/1471-2164-14-501</a>.</p>
</div>
<div id="ref-Topfer2014">
<p>Töpfer, Armin, Tobias Marschall, Rowena A. Bull, Fabio Luciani, Alexander Schönhuth, and Niko Beerenwinkel. 2014. “Viral Quasispecies Assembly via Maximal Clique Enumeration.” Edited by Alice Carolyn McHardy. <em>PLoS Computational Biology</em> 10 (3): e1003515. <a href="https://doi.org/10.1371/journal.pcbi.1003515">https://doi.org/10.1371/journal.pcbi.1003515</a>.</p>
</div>
<div id="ref-Zagordi2011">
<p>Zagordi, Osvaldo, Arnab Bhattacharya, Nicholas Eriksson, and Niko Beerenwinkel. 2011. “ShoRAH: estimating the genetic diversity of a mixed sample from next-generation sequencing data.” <em>BMC Bioinformatics</em> 12 (1): 119. <a href="https://doi.org/10.1186/1471-2105-12-119">https://doi.org/10.1186/1471-2105-12-119</a>.</p>
</div>
<div id="ref-Zagordi2010">
<p>Zagordi, Osvaldo, Lukas Geyrhofer, Volker Roth, and Niko Beerenwinkel. 2010. “Deep sequencing of a genetically heterogeneous sample: local haplotype reconstruction and read error correction.” <em>J. Comput. Biol.</em> 17 (3): 417–28. <a href="https://doi.org/10.1089/cmb.2009.0164">https://doi.org/10.1089/cmb.2009.0164</a>.</p>
</div>
<div id="ref-10.1093/nar/gkq655">
<p>Zagordi, Osvaldo, Rolf Klein, Martin Däumer, and Niko Beerenwinkel. 2010. “Error correction of next-generation sequencing data and reliable estimation of HIV quasispecies.” <em>Nucleic Acids Research</em> 38 (21): 7400–7409. <a href="https://doi.org/10.1093/nar/gkq655">https://doi.org/10.1093/nar/gkq655</a>.</p>
</div>
</div>
