<div id="refs" class="references csl-bib-body hanging-indent"
data-entry-spacing="0" role="list">
<div id="ref-Dreifuss2022.11.02.22281825" class="csl-entry"
role="listitem">
Dreifuss, David, Ivan Topolsky, Pelin Icer Baykal, and Niko
Beerenwinkel. 2022. <span>“Tracking SARS-CoV-2 Genomic Variants in
Wastewater Sequencing Data with LolliPop.”</span> <em>medRxiv</em>. <a
href="https://doi.org/10.1101/2022.11.02.22281825">https://doi.org/10.1101/2022.11.02.22281825</a>.
</div>
<div id="ref-FUHRMANN2021157" class="csl-entry" role="listitem">
Fuhrmann, Lara, Kim Philipp Jablonski, and Niko Beerenwinkel. 2021.
<span>“Quantitative Measures of Within-Host Viral Genetic
Diversity.”</span> <em>Current Opinion in Virology</em> 49: 157–63. <a
href="https://doi.org/10.1016/j.coviro.2021.06.002">https://doi.org/10.1016/j.coviro.2021.06.002</a>.
</div>
<div id="ref-Jahn2021.01.08.21249379" class="csl-entry" role="listitem">
Jahn, Katharina, David Dreifuss, Ivan Topolsky, Anina Kull, Pravin
Ganesanandamoorthy, Xavier Fernandez-Cassi, Carola Bänziger, et al.
2021. <span>“Detection and Surveillance of SARS-CoV-2 Genomic Variants
in Wastewater.”</span> <em>medRxiv</em>. <a
href="https://doi.org/10.1101/2021.01.08.21249379">https://doi.org/10.1101/2021.01.08.21249379</a>.
</div>
<div id="ref-McElroy2013" class="csl-entry" role="listitem">
McElroy, Kerensa, Osvaldo Zagordi, Rowena Bull, Fabio Luciani, and Niko
Beerenwinkel. 2013. <span>“Accurate Single Nucleotide Variant Detection
in Viral Populations by Combining Probabilistic Clustering with a
Statistical Test of Strand Bias.”</span> <em>BMC Genomics</em> 14 (1):
501. <a
href="https://doi.org/10.1186/1471-2164-14-501">https://doi.org/10.1186/1471-2164-14-501</a>.
</div>
<div id="ref-Topfer2014" class="csl-entry" role="listitem">
Töpfer, Armin, Tobias Marschall, Rowena A. Bull, Fabio Luciani,
Alexander Schönhuth, and Niko Beerenwinkel. 2014. <span>“Viral
Quasispecies Assembly via Maximal Clique Enumeration.”</span> Edited by
Alice Carolyn McHardy. <em><span>PLoS</span> Computational Biology</em>
10 (3): e1003515. <a
href="https://doi.org/10.1371/journal.pcbi.1003515">https://doi.org/10.1371/journal.pcbi.1003515</a>.
</div>
<div id="ref-Zagordi2011" class="csl-entry" role="listitem">
Zagordi, Osvaldo, Arnab Bhattacharya, Nicholas Eriksson, and Niko
Beerenwinkel. 2011. <span>“<span class="nocase">ShoRAH: estimating the
genetic diversity of a mixed sample from next-generation sequencing
data.</span>”</span> <em><span>BMC Bioinformatics</span></em> 12 (1):
119. <a
href="https://doi.org/10.1186/1471-2105-12-119">https://doi.org/10.1186/1471-2105-12-119</a>.
</div>
<div id="ref-Zagordi2010" class="csl-entry" role="listitem">
Zagordi, Osvaldo, Lukas Geyrhofer, Volker Roth, and Niko Beerenwinkel.
2010. <span>“<span class="nocase">Deep sequencing of a genetically
heterogeneous sample: local haplotype reconstruction and read error
correction.</span>”</span> <em><span>J. Comput. Biol.</span></em> 17
(3): 417–28. <a
href="https://doi.org/10.1089/cmb.2009.0164">https://doi.org/10.1089/cmb.2009.0164</a>.
</div>
<div id="ref-10.1093/nar/gkq655" class="csl-entry" role="listitem">
Zagordi, Osvaldo, Rolf Klein, Martin Däumer, and Niko Beerenwinkel.
2010. <span>“<span class="nocase">Error correction of next-generation
sequencing data and reliable estimation of HIV
quasispecies</span>.”</span> <em>Nucleic Acids Research</em> 38 (21):
7400–7409. <a
href="https://doi.org/10.1093/nar/gkq655">https://doi.org/10.1093/nar/gkq655</a>.
</div>
</div>
