<!-- markdownlint-disable MD013 -->

# SARS-CoV-2 test data

This data is based on the two synthetic positive controls used as part of the
[SARS-CoV-2 sequencing project done at the BSSE](https://ethz.ch/en/news-and-events/eth-news/news/2020/04/analyses-for-getting-to-grips-with-the-pandemic.html),
and was sequenced at the
[Genomics Facility Basel](https://bsse.ethz.ch/genomicsbasel)
led by Dr. Christian Beisel.

The
[Twist Synthetic SARS-CoV-2 RNA Controls](https://www.twistbioscience.com/resources/twist-synthetic-sars-cov-2-rna-controls)
were diluted 10000 on the plate and used in the initial RT step.

Library preparation adjusted from
[nCoV-2019 sequencing protocol](https://www.protocols.io/view/ncov-2019-sequencing-protocol-bbmuik6w).

A [quick guide to tiling amplicon sequencing and bioinformatics](https://artic.network/quick-guide-to-tiling-amplicon-sequencing-bioinformatics.html).

## Short

- Step 1: reverse transcription, random priming with random hexamers
- Step 2: PCR amplification of viral genome

  [nCoV-2019 PrimalSeq sequencing primers](https://github.com/artic-network/artic-ncov2019/tree/master/primer_schemes/nCoV-2019/V3)
- Step 3: Illumina adapter ligation; canonical Illumina TruSeq adapter sequences

  For [adapter trimming](https://support.illumina.com/bulletins/2016/12/what-sequences-do-i-use-for-adapter-trimming.html?langsel=/us/)

  TruSeq LT and TruSeq HT-based kits:

  - Read 1: AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
  - Read 2: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT

  Trimming should not really be necessary. Amplicons 400ish bp. Read length 251 bases

## Sequencing

[MiSeq](https://emea.illumina.com/systems/sequencing-platforms/miseq.html)
PE 2x251

(v3 600 cycles kit)

## Testdata packaging

The REF_aln.bam files generated by [V-pipe](https://cbg-ethz.github.io/V-pipe/sars-cov-2/)
(as part of the
[SARS-CoV-2 sequencing project](https://bsse.ethz.ch/cevo/cevo-press/2020/05/first-data-for-genomic-surveillance-of-sars-cov-2-in-switzerland-made-available.html))
have been:

- [capped to 512](https://www.biostars.org/p/154220/)
  using [jvarkit](https://lindenb.github.io/jvarkit/Biostar154220.html),
- [converted back to .fastq raw_reads](http://www.metagenomics.wiki/tools/samtools/converting-bam-to-fastq)
  using [`samtools bam2fq`](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/samtools/bam2fq/separate.html)
- and finally GZ-compressed using
  [`advdef --shrink-insane`](https://www.advancemame.it/comp-readme.html).
